### ISSUE TITLE
kubernetes_manifest does not wait for dependencies

### ISSUE THREAD

**Issue by iosifnicolae2** [hashicorp/terraform-provider-kubernetes#1782]
Kubernetes_manifest does not wait for dependencies.

### Terraform Version, Provider Version and Kubernetes Version
```
Terraform version:  v1.2.4
Kubernetes provider version: v2.6.1
Kubernetes version:  v1.21.5+k3s2
```

### Affected Resource(s)
- kubernetes_manifest

### Terraform Configuration Files
```hcl
resource "kubernetes_manifest" "longhorn_backups" {
  depends_on = [rancher2_app_v2.rancher_longhorn]
  manifest   = {
    "apiVersion" : "longhorn.io/v1beta1",
    "kind" : "RecurringJob",
    "metadata" : {
      "name" : "automatic-backups",
      "namespace" : "longhorn-system"
    },
    "spec" : {
      "concurrency" : 3,
      "cron" : var.longhorn_backup_cron,
      "retain" : 1,
      "task" : "backup"
      "groups" : ["default"]
    }
  }
}
```

### Panic Output
The plugin returned an unexpected error from plugin.(*GRPCProvider).ReadResource: rpc error: code = Unknown desc = no matches for kind "RecurringJob" in version "longhorn.io/v1beta1"

### Steps to Reproduce
1. terraform apply

### Expected Behavior
We were expecting the kubernetes_manifest plugin to wait on dependencies and then check the status of the manifest.

### Actual Behavior
We're getting an error because the CRD was not deployed

**Comment by steve-gray** [comment]
What advantage does validating the manifest during the apply phase have? I'm at a loss here as to the rationale between the limitation/constraint on the usage of this and the down-wind complexity this decision brings to people using this module. The module would work equally effectively if the CRD was validated at the apply stage, because you'd need all the variable values interpolated from the spec to even know if they're valid - so it has to validate again later anyway.

CRD's are inherently a runtime consideration, dynamically updatable on the fly in the cluster, and something where static planning of them during the plan phase offers nothing? Helm and other providers don't error in similar ways and preventing people being able to deploy a cluster and the dependant resources in a single pass feels like a bad call.

**Comment by pathob** [comment]
@alexsomesan 

I have to agree on everything that @steve-gray is saying.

> The limitation stems from the way Terraform itself arranges the operations of planning and applying. All resources that are part of the same operation are first planned for and only after all resources are finished planning do they begin to get applied.

I don't see a contradiction here to what @steve-gray is suggesting. Why can't we plan a CR based on a CRD that doesn't exist yet? Terraform doesn't have to be that strict and also isn't in practice in so many cases. I'm mostly working with the `azurerm` provider and how many times have I seen Terraform failing during the apply step for ridiculous reasons like a case-sensitive string validation that has not been done during the plan step. So why be so strict here?

Another approach that would match the way for example Helm works is to add a `validate_during_plan` flag or similarly named, just as `helm template` does, see https://helm.sh/docs/helm/helm_template/

> There is a simple work-around to this issue: split your configuration into two (or more) applies. The first apply should create all the CRDs, then you can move on to a second apply operation to create the CR objects and any other K8s resources you may need.

That's not as simple as it sounds like. Many workflows just don't work like that. We're for example using Terragrunt and its mocking features to run ALL plans first and then ALL applies. So splitting this up into different modules wouldn't solve the issue.

I would like to see this discussion reopened. Thanks in advance.

**Comment by kutysam** [comment]
> What advantage does validating the manifest during the apply phase have? I'm at a loss here as to the rationale between the limitation/constraint on the usage of this and the down-wind complexity this decision brings to people using this provider. The provider would work equally effectively if the CRD was validated at the apply stage, because you'd need all the variable values interpolated from the spec to even know if they're valid - so it has to validate again later anyway (i.e. you're already on the wrong side of the airtight hatch-way when it comes to spec validation).
> 
> CRD's are inherently a runtime consideration, dynamically updatable on the fly in the cluster, and something where static planning of them during the plan phase offers nothing? Helm and other providers don't error in similar ways and preventing people being able to deploy a cluster and the dependant resources in a single pass feels like a bad call.
> 
> I'm happy to do up a PR to give this the punt to the right location if folks buy into it?

Assuming you are creating argocd helm_release and then running a manifest 
```
    apiVersion = "argoproj.io/v1alpha1"
    kind       = "Application"
```

It will not work because helm_release needs to be run and depends_on for kubernetes manifest will not work

**Comment by observe-ryan** [comment]
I agree with @steve-gray and @pathob here; failing because you can't validate that the manifest will apply correctly because the control plane isn't available is a bug. In my opinion, the situation where the api server cannot be contacted should not fail the plan or apply but rather put those resources into a "known after apply" stage.

At this point, a plan will "succeed" and an apply will handle any upcoming error as the resources are actually applied. If the appropriate `depends_on` isn't defined then the apply should fail with the appropriate error code.

**Comment by online01993** [comment]
up, pls!

**Comment by suryastef** [comment]
there is a workaround for this by using separate kubernetes manifest (yaml) file `resource "kubectl_manifest"` from [gavinbunney/kubectl](https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/resources/kubectl_manifest), and of course you can't manipulate the manifest (yaml) file using tf var, locals, etc

**Comment by PietervdWerk** [comment]
@suryastef kubectl_manifest workaround works, though for a reason I cannot explain the error showed now after the resource was applied (while the resources applied successfully). I added `wait = true` which resolved this issue.

>[wait](https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/resources/kubectl_manifest#wait) - Optional. Set this flag to wait or not for finalized to complete for deleted objects. Default false.

**Comment by rayrapetyan** [comment]
Is there any other workaround than putting:
```
count = var.create_istio_vs == "true" ? 1 : 0
```
into every manifest and run terraform apply twice like this:
```
terraform apply
terraform apply -var "create_istio_vs=true" -auto-approve
```
?
The workaround above works fine for me, but just curious if there are more.

**Comment by MB175** [comment]
I would actually suggest to not only have a validate boolean which we can trigger by certain expressions, but also have a boolean to trigger skip validation if dependency not satisfied.

**Comment by mwmahlberg** [comment]
> There is a simple work-around to this issue: split your configuration into two (or more) applies.

I would not exactly call that easy, because it is not that cleanly cut. A simple example: A helm chart provides the CRDs for kind foo. A Foo needs to be configured for Helm Chart "bar" to properly work (secrets provided via vso comes to my mind here as an example), though helm chart bar itself provides CRDs. This leads to a situation where you basically have multiple applies, potentially ad infinitum, to roll out anything but the most basic infrastructure.

Personally, I'd say that if the CRD is not available for any reason (say for example the k8s cluster is supposed to be created within the same configuration), it is safe to assume that it has to be applied, because if the CRD still is not present after the dependencies are rolled out, an error will occur either way.

**Comment by atimofeev** [comment]
This is still actual. Fully bootstrapping the cluster with terraform is a pain due to issues like this one.

**Comment by bobbyiliev** [comment]
This is also a blocker for me too for a couple of projects.

**Comment by aleprice** [comment]
It continues to be a pain.
I'm installing a Velero "helm_release" and a schedule for backups with "kubernetes_manifest", and I'm having the same issue.

```
Error: API did not recognize GroupVersionKind from manifest (CRD may not be installed)

  on ../../modules/velero/backup.tf line 1, in resource "kubernetes_manifest" "velero_schedule":
   1: resource "kubernetes_manifest" "velero_schedule" {

no matches for kind "Schedule" in group "velero.io"
```
